pkg/proxy/endpointslicecache.go

```diff
@@ -0,0 +1,190 @@
+		Endpoints: []EndpointInfo{},
+	}
+	for _, endpoint := range endpointSlice.Endpoints {
+		if endpoint.Conditions.Ready == nil || *endpoint.Conditions.Ready == true {
+			esInfo.Endpoints = append(esInfo.Endpoints, EndpointInfo{
+				Addresses: endpoint.Addresses,
+				Topology:  endpoint.Topology,
+			})
+		}
+	}
+	if _, exists := e.slicesByService[serviceKey]; !exists {
+		e.slicesByService[serviceKey] = map[string]EndpointSliceInfo{}
+	}
+	e.slicesByService[serviceKey][sliceKey] = esInfo
+}
+
+// Delete a slice from the cache
+func (e *EndpointSliceCache) Delete(endpointSlice *discovery.EndpointSlice) {
+	serviceKey, sliceKey := EndpointSliceCacheKeys(endpointSlice)
+	delete(e.slicesByService[serviceKey], sliceKey)
+}
+
+// GetEndpointsMap computes an EndpointsMap for a given service
+func (e *EndpointSliceCache) GetEndpointsMap(serviceNN types.NamespacedName) EndpointsMap {
+	endpointsMap := EndpointsMap{}
+	endpointInfoByServicePort := map[ServicePortName]map[string]Endpoint{}
+	endpointSlices := e.slicesByService[serviceNN]
+
+	// We need to build a map of portname -> all ip:ports for that portname.
+	for _, endpointSlice := range endpointSlices {
+		for i := range endpointSlice.Ports {
+			port := &endpointSlice.Ports[i]
+			if port.Port == nil || *port.Port == int32(0) {
+				klog.Warningf("ignoring invalid endpoint port %s", *port.Name)
+				continue
+			}
+
+			svcPortName := ServicePortName{NamespacedName: serviceNN}
+			if port.Name != nil {
+				svcPortName.Port = *port.Name
+			}
+			if _, exists := endpointInfoByServicePort[svcPortName]; !exists {
+				endpointInfoByServicePort[svcPortName] = map[string]Endpoint{}
+			}
+
+			// iterate through endpoints to add them to endpointInfoByServicePort
+			for i := range endpointSlice.Endpoints {
+				endpoint := &endpointSlice.Endpoints[i]
+				if len(endpoint.Addresses) < 1 {
+					klog.Warningf("ignoring invalid endpoint port %s with empty host", *port.Name)
+					continue
+				}
+
+				// Filter out the incorrect IP version case.
+				// Any endpoint port that contains incorrect IP version will be ignored.
+				if e.isIPv6Mode != nil && utilnet.IsIPv6String(endpoint.Addresses[0]) != *e.isIPv6Mode {
+					// Emit event on the corresponding service which had a different
+					// IP version than the endpoint.
+					utilproxy.LogAndEmitIncorrectIPVersionEvent(e.recorder, "endpointslice", endpoint.Addresses[0], serviceNN.Name, serviceNN.Namespace, "")
+					continue
+				}
+
+				isLocal := len(e.hostname) > 0 && endpoint.Topology["kubernetes.io/hostname"] == e.hostname
+				endpointInfo := newBaseEndpointInfo(endpoint.Addresses[0], int(*port.Port), isLocal)
+
+				// This logic ensures we're deduping potential overlapping endpoints
+				// isLocal should not vary between matching IPs, but if it does, we favor a true value here if it exists
+				if _, exists := endpointInfoByServicePort[svcPortName][endpointInfo.IP()]; !exists || isLocal {
+					if e.makeEndpointInfo != nil {
```

Comment:
we should just initialize makeEndpointInfo (mutateEndpointInfo to a pass through and get rid of all of these branches if the func is nil

Reply:
good call, done